README.md

```diff
@@ -138,6 +143,65 @@ and then propogate or discard the resulting
 `TypeError: descriptor 'isdigit' for 'str' objects doesn't apply to a 'int' object`
 exception.  We encourage libraries to document the behaviour they choose.
 
+### Consuming and creating metadata
+
+#### `BaseMetadata`
+
+We provide `BaseMetadata` as base class for all metadata that serves as a namespace and identifier to easily filter metadata from this package out from other metadata that may be contained in the `Annotated[...]` type annotation.
+
+If you are creating your own metadata, you _should_ inherit from `annotated_types.BaseMetadata`:
```

Comment:
I think I disagree here - for example if Pydantic has some metadata inheriting from BaseMetadata, that doesn't help Hypothesis resolve it at all; we'd still need specific knowledge.

(on the other hand inheriting from GroupedMetadata *does* help, precisely because it tells us to iterate over it and pluck out each BaseMetadata)

IMO the reason to have a BaseMetadata class at all is so that Hypothesis can tell at runtime when it's been passed things from a newer version of annotated-types than it understands, and request that the user update *Hypothesis*.

Reply:
I actually agree with you! I think I just worded poorly. In my mind a `GroupedMetadata` is to an `int` how a `List[int]` is to an `int`: it is not an `int` in and of itself. In other words, `GroupedMetada` _is not metadata_. Hence in my mind my wording of "If you are creating your own metadata, you _should_ inherit from `annotated_types.BaseMetadata`" made sense because if you are creating a `Field`.

I am also not saying that inheriting from `BaseMetadata` automatically makes consumers like Hypothesis be able to process this metadata. Instead, if lets Hypothesis identify that this is metadata and decide what to do with it (if Hypothesis chooses to ignore unknown subclasses of `BaseMetadata` or raise an error if it encounters them is up to Hypothesis).

Reply:
> "If you are creating your own metadata, you _should_ inherit from `annotated_types.BaseMetadata`"

I meant that nobody outside the `annotated-types` package itself should inherit from `BaseMetadata`, because doing so means tools (e.g. Hypothesis) lose the ability to distinguish between the user installing a tool older than the current annotated-types version, vs the user installing something that also provides metadata types.

To me "noticing when there's metadata that I can't handle but expect to be able to" is the *only* sensible use of `BaseMetadata`; otherwise I'd have our metadata classes inherit directly from `object.

Put another way, I'd like to error on unknown metadata *if and only if* they're supplied by the `annotated-types` pacakge.

Reply:
> Put another way, I'd like to error on unknown metadata if and only if they're supplied by the annotated-types pacakge.

I see. My thought was that the tool (Hypothesis) in this case would error or ignore (up to the implementer) _any_ metadata it does not recognize, from `annotated-types` or not. If we want _different_ behavior for `annotated-types` vs. other user installed packages we'd need to add a level to the inheritance chain:

1. `BaseMetadata` namespaces anything in the `annotated-metadata` ecosystem from arbitrary stuff in `Annotated` (`Annotated[..., BaseMetadata(), "foo"]`).
2. `AnnotatedTypesMetadata` (or something like this) which namespaces all of the metadata provided by `annotated-types` itself
3. `Gt`, etc. are the concrete constraints.

```python
from annotated_types import Len, BaseMetadata, AnnotatedTypesMetadata

class PydanticPattern(BaseMetadata):
    ...

SomeField = Annotated[str, Len(1), PydanticPattern(...)]

# in some parser
args = list(get_args(SomeField))[1:]
for arg in args:
    if not isinstace(arg, BaseMetadata):
        continue
    metadata_type = type(arg)
    if metadata_type in KNOWN_METADATA_TYPES:
        ...  # do something
    # we don't know this metadata, up to the implementer to decide what to do
    # hypothesis could do the following
    if isinstance(arg, AnnotatedTypesMetadata):
        # unknown metadata coming from annotated-types
        raise ValueError("Please update hypothesis")
    # this is an unsupported package
    raise ValueError(f"Metadata of type {metadata_type} is not supported")
```

Reply:
Alternatively we could avoid the extra level of inheritance if Hypothesis is willing to check the module (`if type(arg).__module__.startswith("annotated_types")` or something like this).

Reply:
But why does Pydantic want to inherit from `BaseMetadata`?  The class doesn't actually do anything, and there's nothing that downstream libraries can do with that information if they don't have special knowledge of Pydantic anyway!

Reply:
I think it would allow writing "generic" functions for the `annotated-types` "ecosystem" that do stuff like parse a type annotation into a list of metadata, which Hypothesis can then individually look at and consume, ignore, error, warn, or whatever it wants to do. Without that base class the only way to differentiate a piece of metadata is by exhaustive checking, which makes it harder to write generic utility functions because you have to pass around some enumeration of all known metadata classes as well as instructions as to what to do for stuff that isn't in that enumeration. I say this having written these sorts of parsers for `annotated-types` twice (once in our tests here and once in https://github.com/adriangb/typeval).

Reply:
And I think this generic functionality is important to enable building an ecosystem and not just a collection of metadata ( I realize we haven't really formally decided which one we are yet)

Reply:
My understanding was that [we're right up the minimal-collection-of-metadata end *in this package*](https://github.com/annotated-types/annotated-types#design--history):

> ### Design & History
> This package was designed at the PyCon 2022 sprints by the maintainers of Pydantic and Hypothesis, with the goal of making it as easy as possible for end-users to provide more informative annotations for use by runtime libraries.
>
> **It is deliberately minimal**, and following PEP-593 allows considerable downstream discretion in what (if anything!) they choose to support. Nonetheless, we expect that staying simple and covering only the most common use-cases will give users and maintainers the best experience we can. If you'd like more constraints for your types - follow our lead, by defining them and documenting them downstream!

(emphasis added)